File: Árvore de Decisão IA/ÁrvoreDeDecisão.R
# ── Instalar pacotes (execute uma vez) ───────────────────────
# install.packages(c("rpart", "rpart.plot", "caret", "pROC", "ggplot2", "dplyr"))

library(rpart)       
library(rpart.plot)  
library(caret)       
library(pROC)        
library(ggplot2)
library(dplyr)

set.seed(42)


cat("======================================================\n")
cat("INSPEÇÃO INICIAL\n")
cat("======================================================\n")

df <- read.csv('...data/UCI_Credit_Card.csv', stringsAsFactors = FALSE)

cat("Dimensões:", dim(df), "\n")
cat("\nColunas:\n"); print(names(df))
cat("\nDistribuição da variável alvo:\n")
print(table(df$default.payment.next.month))
cat("\nProporção (%):\n")
print(round(prop.table(table(df$default.payment.next.month)) * 100, 2))


cat("\n======================================================\n")
cat("PRÉ-PROCESSAMENTO\n")
cat("======================================================\n")

df$ID <- NULL

names(df)[names(df) == "default.payment.next.month"] <- "default"

df$default <- factor(df$default,
                     levels = c(0, 1),
                     labels = c("NaoInadim", "Inadim"))

cat("Valores ausentes:", sum(is.na(df)), "\n")
cat("Observações finais:", nrow(df), "\n")
cat("Atributos:", ncol(df) - 1, "\n")


idx    <- createDataPartition(df$default, p = 0.80, list = FALSE)
treino <- df[ idx, ]
teste  <- df[-idx, ]

cat(sprintf("\nTreino : %d registros\n", nrow(treino)))
cat(sprintf("Teste  : %d registros\n",  nrow(teste)))


cat("\n======================================================\n")
cat("TREINAMENTO — ÁRVORE DE DECISÃO\n")
cat("======================================================\n")

arvore <- rpart(
  default ~ .,
  data    = treino,
  method  = "class",
  parms   = list(split = "gini"),       
  control = rpart.control(
    cp        = 0.001,   
    maxdepth  = 5,       
    minsplit  = 100,     
    minbucket = 50       
  )
)

cp_otimo <- arvore$cptable[which.min(arvore$cptable[, "xerror"]), "CP"]
cat(sprintf("CP ótimo (menor xerror): %.6f\n", cp_otimo))

arvore_podada <- prune(arvore, cp = cp_otimo)

cat(sprintf("Número de nós terminais (folhas): %d\n",
            sum(arvore_podada$frame$var == "<leaf>")))


cat("\n======================================================\n")
cat("AVALIAÇÃO NO CONJUNTO DE TESTE\n")
cat("======================================================\n")

y_pred <- predict(arvore_podada, teste, type = "class")
y_prob <- predict(arvore_podada, teste, type = "prob")[, "Inadim"]
y_real <- teste$default

cm  <- confusionMatrix(y_pred, y_real, positive = "Inadim")
acc  <- cm$overall["Accuracy"]
prec <- cm$byClass["Precision"]
rec  <- cm$byClass["Recall"]
f1   <- cm$byClass["F1"]

roc_obj <- roc(y_real, y_prob, quiet = TRUE)
auc_val <- as.numeric(auc(roc_obj))

cat(sprintf("\n  Acurácia:  %.4f (%.2f%%)\n", acc, acc * 100))
cat(sprintf("  Precisão:  %.4f\n", prec))
cat(sprintf("  Revocação: %.4f\n", rec))
cat(sprintf("  F1-score:  %.4f\n", f1))
cat(sprintf("  AUC-ROC:   %.4f\n", auc_val))
cat("\n  Matriz de Confusão:\n")
print(cm$table)
cat("\n  Relatório completo:\n")
print(cm)

ctrl   <- trainControl(method = "cv", number = 5,
                       classProbs = TRUE,
                       summaryFunction = twoClassSummary)
cv_mod <- train(default ~ ., data = df, method = "rpart",
                trControl = ctrl, metric = "ROC",
                tuneLength = 10)
cat(sprintf("\n  AUC médio (5-fold CV): %.4f\n", max(cv_mod$results$ROC)))


cat("\n======================================================\n")
cat("IMPORTÂNCIA DAS VARIÁVEIS\n")
cat("======================================================\n")

imp <- arvore_podada$variable.importance
if (length(imp) > 0) {
  imp_df <- data.frame(
    variavel    = names(imp),
    importancia = as.numeric(imp)
  ) %>% arrange(desc(importancia))
  print(imp_df)
} else {
  cat("Nenhuma variável com importância registrada.\n")
}


png("arvore_decisao_R.png", width = 1400, height = 700, res = 120)
rpart.plot(
  arvore_podada,
  type    = 4,       
  extra   = 104,     
  fallen.leaves = TRUE,
  main    = "Árvore de Decisão (podada) — UCI Credit Card (R)",
  cex     = 0.75
)
dev.off()
cat("\n✔ arvore_decisao_R.png salvo.\n")

if (length(imp) > 0) {
  top_imp <- head(imp_df, 10)
  top_imp$variavel <- factor(top_imp$variavel,
                             levels = top_imp$variavel[order(top_imp$importancia)])
  p <- ggplot(top_imp, aes(x = variavel, y = importancia)) +
    geom_col(fill = "steelblue") +
    coord_flip() +
    labs(title = "Top Variáveis — Árvore de Decisão (R)",
         x = NULL, y = "Importância") +
    theme_minimal(base_size = 13)
  ggsave("importancia_dt_R.png", p, width = 8, height = 5, dpi = 150)
  cat("✔ importancia_dt_R.png salvo.\n")
}

png("roc_dt_R.png", width = 700, height = 600, res = 120)
plot(roc_obj,
     main = sprintf("Curva ROC — Árvore de Decisão (R)  AUC = %.3f", auc_val),
     col  = "steelblue", lwd = 2,
     xlab = "Taxa de Falso Positivo",
     ylab = "Taxa de Verdadeiro Positivo (Revocação)")
abline(0, 1, lty = 2, col = "gray60")
dev.off()
cat("✔ roc_dt_R.png salvo.\n")

png("cp_table_R.png", width = 700, height = 500, res = 120)
plotcp(arvore, main = "Complexidade vs. Erro — Árvore de Decisão (R)")
dev.off()
cat("✔ cp_table_R.png salvo.\n")

cat("\n✔ Execução concluída com sucesso.\n")

File: Árvore de Decisão IA/ÁrvoreDeDecisão.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.tree import DecisionTreeClassifier, plot_tree, export_text
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import (
    accuracy_score, precision_score, recall_score,
    f1_score, confusion_matrix, classification_report,
    roc_auc_score, roc_curve
)

SEED = 42
np.random.seed(SEED)

df = pd.read_csv('...data/UCI_Credit_Card.csv')

print("=" * 60)
print("INSPEÇÃO INICIAL")
print("=" * 60)
print(f"Dimensões: {df.shape}")
print(f"\nPrimeiras linhas:\n{df.head(3)}")
print(f"\nDistribuição da variável alvo:")
print(df["default.payment.next.month"].value_counts())
print(f"\nProporção (%):")
print((df["default.payment.next.month"].value_counts(normalize=True) * 100).round(2))


print("\n" + "=" * 60)
print("PRÉ-PROCESSAMENTO")
print("=" * 60)

df.drop(columns=["ID"], inplace=True)

print(f"Valores ausentes: {df.isnull().sum().sum()}")

df.rename(columns={"default.payment.next.month": "default"}, inplace=True)

X = df.drop(columns=["default"])
y = df["default"]

print(f"\nAtributos: {X.shape[1]} variáveis")
print(f"Observações: {len(y)}")


X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.20, random_state=SEED, stratify=y
)

print(f"\nTreino : {X_train.shape[0]} registros")
print(f"Teste  : {X_test.shape[0]} registros")


print("\n" + "=" * 60)
print("TREINAMENTO — ÁRVORE DE DECISÃO")
print("=" * 60)

dt = DecisionTreeClassifier(
    criterion        = "gini",   
    max_depth        = 5,        
    min_samples_leaf = 50,       
    min_samples_split= 100,      
    random_state     = SEED
)

dt.fit(X_train, y_train)

print(f"Profundidade real da árvore: {dt.get_depth()}")
print(f"Número de folhas           : {dt.get_n_leaves()}")


print("\n" + "=" * 60)
print("AVALIAÇÃO NO CONJUNTO DE TESTE")
print("=" * 60)

y_pred = dt.predict(X_test)
y_prob = dt.predict_proba(X_test)[:, 1]

acc  = accuracy_score(y_test, y_pred)
prec = precision_score(y_test, y_pred, zero_division=0)
rec  = recall_score(y_test, y_pred, zero_division=0)
f1   = f1_score(y_test, y_pred, zero_division=0)
auc  = roc_auc_score(y_test, y_prob)
cm   = confusion_matrix(y_test, y_pred)

print(f"\n  Acurácia:  {acc:.4f}  ({acc*100:.2f}%)")
print(f"  Precisão:  {prec:.4f}")
print(f"  Revocação: {rec:.4f}")
print(f"  F1-score:  {f1:.4f}")
print(f"  AUC-ROC:   {auc:.4f}")

print(f"\n  Matriz de Confusão:")
print(f"  {cm}")

print(f"\n  Relatório Completo:")
print(classification_report(y_test, y_pred,
      target_names=["Não Inadimplente", "Inadimplente"]))

cv_f1 = cross_val_score(dt, X, y, cv=5, scoring="f1")
cv_auc = cross_val_score(dt, X, y, cv=5, scoring="roc_auc")
print(f"  F1  médio 5-fold CV: {cv_f1.mean():.4f} ± {cv_f1.std():.4f}")
print(f"  AUC médio 5-fold CV: {cv_auc.mean():.4f} ± {cv_auc.std():.4f}")


print("\n" + "=" * 60)
print("IMPORTÂNCIA DAS VARIÁVEIS (Gini)")
print("=" * 60)

importancias = (
    pd.Series(dt.feature_importances_, index=X.columns)
    .sort_values(ascending=False)
)
print(importancias[importancias > 0].to_string())



fig, ax = plt.subplots(figsize=(5, 4))
sns.heatmap(cm, annot=True, fmt="d", cmap="Blues", ax=ax,
            xticklabels=["Não Inadimplente", "Inadimplente"],
            yticklabels=["Não Inadimplente", "Inadimplente"])
ax.set_title("Matriz de Confusão — Árvore de Decisão (Python)")
ax.set_xlabel("Previsto")
ax.set_ylabel("Real")
plt.tight_layout()
plt.savefig("matriz_confusao_dt_python.png", dpi=150)
plt.close()
print("\n✔ matriz_confusao_dt_python.png salvo.")

top10 = importancias[importancias > 0].head(10)
fig, ax = plt.subplots(figsize=(8, 5))
top10.sort_values().plot(kind="barh", color="steelblue", ax=ax)
ax.set_title("Top Variáveis — Árvore de Decisão (Python)")
ax.set_xlabel("Importância (Gini)")
plt.tight_layout()
plt.savefig("importancia_dt_python.png", dpi=150)
plt.close()
print("✔ importancia_dt_python.png salvo.")

fpr, tpr, _ = roc_curve(y_test, y_prob)
fig, ax = plt.subplots(figsize=(6, 5))
ax.plot(fpr, tpr, label=f"Árvore de Decisão (AUC = {auc:.3f})", color="steelblue")
ax.plot([0, 1], [0, 1], "k--", label="Aleatório")
ax.set_xlabel("Taxa de Falso Positivo")
ax.set_ylabel("Taxa de Verdadeiro Positivo (Revocação)")
ax.set_title("Curva ROC — Árvore de Decisão (Python)")
ax.legend()
plt.tight_layout()
plt.savefig("roc_dt_python.png", dpi=150)
plt.close()
print("✔ roc_dt_python.png salvo.")

fig, ax = plt.subplots(figsize=(20, 8))
plot_tree(dt, feature_names=X.columns.tolist(),
          class_names=["Não Inadim.", "Inadim."],
          filled=True, rounded=True, fontsize=9, ax=ax)
ax.set_title("Árvore de Decisão (profundidade=5) — UCI Credit Card", fontsize=13)
plt.tight_layout()
plt.savefig("arvore_decisao_python.png", dpi=120)
plt.close()
print("✔ arvore_decisao_python.png salvo.")

print("\n" + "=" * 60)
print("REGRAS DA ÁRVORE (texto)")
print("=" * 60)
print(export_text(dt, feature_names=list(X.columns), max_depth=3))

print("\n✔ Execução concluída com sucesso.")